af_ktls.c

```diff
@@ -901,20 +960,50 @@ static int tls_sendmsg(struct socket *sock, struct msghdr *msg, size_t size)
 	sg_unmark_end(tsk->sgl_send[cnt-1].sg + tsk->sgl_send[cnt-1].npages - 1);
 	sg_chain(tsk->sgl_send[cnt-1].sg, tsk->sgl_send[cnt-1].npages + 1, tsk->sgtag_send);
 
+	npages = tls_pre_encrypt(tsk, size);
+	if (npages < 0) {
+		ret = npages;
+		goto send_end;
+	}
+
 	ret = tls_do_encryption(tsk, tsk->sgaad_send, tsk->sg_tx_data, size);
 	if (ret < 0)
-		goto send_end;
+		goto encrypt_failed;
 
 	tls_make_prepend(tsk, tsk->header_send, size);
-
-	ret = kernel_sendmsg(tsk->socket, msg, tsk->vec_send, KTLS_VEC_SIZE,
-			KTLS_RECORD_SIZE(tsk, size));
+	memcpy(page_address(tsk->pages_send), tsk->header_send, KTLS_PREPEND_SIZE(tsk));
```

Comment:
Needs rebase and:
```diff
diff --git a/af_ktls.c b/af_ktls.c
index 55aab06..7e43265 100644
--- a/af_ktls.c
+++ b/af_ktls.c
@@ -970,8 +970,7 @@ static int tls_sendmsg(struct socket *sock, struct msghdr *msg, size_t size)
 	if (ret < 0)
 		goto encrypt_failed;
 
-	tls_make_prepend(tsk, tsk->header_send, size);
-	memcpy(page_address(tsk->pages_send), tsk->header_send, KTLS_PREPEND_SIZE(tsk));
+	tls_make_prepend(tsk, page_address(tsk->pages_send), size);
 	/*
 	 * The whole record is sent in one go, instead of 3 separate calls
 	 * to kernel_sendpage. We don't want to deal with the situation
```